App crashes when tracking DerivedState

Reproduce Step : Just launch app with this plugin applied.
composeBomVersion = "2023.10.01"
composeCompiler = "1.5.10"

Stacktrace: 

`java.lang.RuntimeException: Unable to resume activity {com.gangnam.sister.debug/com.kotlin.activity.MainActivity}: java.lang.ClassCastException: androidx.compose.runtime.DerivedSnapshotState cannot be cast to androidx.compose.runtime.snapshots.SnapshotMutableState
                                                                                                    	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:5426)
                                                                                                    	at android.app.ActivityThread.handleResumeActivity(ActivityThread.java:5507)
                                                                                                    	at android.app.servertransaction.ResumeActivityItem.execute(ResumeActivityItem.java:57)
                                                                                                    	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.executeLifecycleState(TransactionExecutor.java:180)
                                                                                                    	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:98)
                                                                                                    	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2685)
                                                                                                    	at android.os.Handler.dispatchMessage(Handler.java:106)
                                                                                                    	at android.os.Looper.loopOnce(Looper.java:230)
                                                                                                    	at android.os.Looper.loop(Looper.java:319)
                                                                                                    	at android.app.ActivityThread.main(ActivityThread.java:8913)
                                                                                                    	at java.lang.reflect.Method.invoke(Native Method)
                                                                                                    	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:608)
                                                                                                    	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1103)
                                                                                                    Caused by: java.lang.ClassCastException: androidx.compose.runtime.DerivedSnapshotState cannot be cast to androidx.compose.runtime.snapshots.SnapshotMutableState
                                                                                                    	at land.sungbin.composeinvestigator.runtime.ComposeStateObjectValueGetter.getCurrentValue(StateObjectTracker.kt:234)
                                                                                                    	at land.sungbin.composeinvestigator.runtime.ComposeStateObjectValueGetter.initialize$runtime_release(StateObjectTracker.kt:240)
                                                                                                    	at land.sungbin.composeinvestigator.runtime.StateObjectTrackerKt$registerStateObjectTracking$1$register$2$1.onRemembered(StateObjectTracker.kt:159)
                                                                                                    	at androidx.compose.runtime.CompositionImpl$RememberEventDispatcher.dispatchRememberObservers(Composition.kt:1295)
                                                                                                    	at androidx.compose.runtime.CompositionImpl.applyChangesInLocked(Composition.kt:984)
                                                                                                    	at androidx.compose.runtime.CompositionImpl.applyChanges(Composition.kt:1005)
                                                                                                    	at androidx.compose.runtime.Recomposer.composeInitial$runtime_release(Recomposer.kt:1099)
                                                                                                    	at androidx.compose.runtime.CompositionImpl.composeInitial(Composition.kt:633)
                                                                                                    	at androidx.compose.runtime.CompositionImpl.setContent(Composition.kt:619)
                                                                                                    	at androidx.compose.ui.platform.WrappedComposition$setContent$1.invoke(Wrapper.android.kt:123)
                                                                                                    	at androidx.compose.ui.platform.WrappedComposition$setContent$1.invoke(Wrapper.android.kt:114)
                                                                                                    	at androidx.compose.ui.platform.AndroidComposeView.setOnViewTreeOwnersAvailable(AndroidComposeView.android.kt:1289)
                                                                                                    	at androidx.compose.ui.platform.WrappedComposition.setContent(Wrapper.android.kt:114)
                                                                                                    	at androidx.compose.ui.platform.WrappedComposition.onStateChanged(Wrapper.android.kt:164)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry$ObserverWithState.dispatchEvent(LifecycleRegistry.kt:314)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.forwardPass(LifecycleRegistry.kt:251)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.sync(LifecycleRegistry.kt:287)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.moveToState(LifecycleRegistry.kt:136)
                                                                                                    	at androidx.lifecycle.LifecycleRegistry.handleLifecycleEvent(LifecycleRegistry.kt:119)
                                                                                                    	at androidx.fragment.app.FragmentViewLifecycleOwner.handleLifecycleEvent(FragmentViewLifecycleOwner.java:100)
                                                                                                    	at androidx.fragment.app.Fragment.restoreViewState(Fragment.java:708)
                                                                                                    	at androidx.fragment.app.Fragment.restoreViewState(Fragment.java:3176)
                                                                                                    	at androidx.fragment.app.Fragment.performActivityCreated(Fragment.java:3161)
                                                                                                    	at androidx.fragment.app.FragmentStateManager.activityCreated(FragmentStateManager.java:619)
                                                                                                    	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:275)
                                                                                                    	at androidx.fragment.app.FragmentManager.executeOpsTogether(FragmentManager.java:1943)
2024-03-05 13:21:40.287 26853-26853 AndroidRuntime          com.gangnam.sister.debug             E  	at androidx.fragment.app.FragmentManager.removeRedundantOperationsAndExecute(FragmentManager.java:1839)
                                                                                                    	at androidx.fragment.app.FragmentManager.execSingleAction(FragmentManager.java:1751)
                                                                                                    	at androidx.fragment.app.BackStackRecord.commitNowAllowingStateLoss(BackStackRecord.java:323)
                                                                                                    	at androidx.fragment.app.FragmentPagerAdapter.finishUpdate(FragmentPagerAdapter.java:249)
                                                                                                    	at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1244)
                                                                                                    	at androidx.viewpager.widget.ViewPager.populate(ViewPager.java:1092)
                                                                                                    	at androidx.viewpager.widget.ViewPager.setAdapter(ViewPager.java:540)
                                                                                                    	at com.kotlin.activity.MainActivity.initAdapter(MainActivity.kt:601)
                                                                                                    	at com.kotlin.presenter.MainPresenter.onResume(MainPresenter.kt:78)
                                                                                                    	at com.kotlin.activity.MainActivity.onResume(MainActivity.kt:561)
                                                                                                    	at android.app.Instrumentation.callActivityOnResume(Instrumentation.java:1603)
                                                                                                    	at android.app.Activity.performResume(Activity.java:9119)
                                                                                                    	at android.app.ActivityThread.performResumeActivity(ActivityThread.java:5399)
                                                                                                    	... 13 more`

--- a/documentation/docs/changelog.md
+++ b/documentation/docs/changelog.md
@@ -1,11 +1,19 @@
 # Change Log
 
-ComposeInvestigator is heavily dependent on the version of [Compose Compiler](https://developer.android.com/jetpack/androidx/releases/compose-compiler),
-so the version of ComposeInvestigator follows the format `[Compose Compiler Version - ComposeInvestigator Version]`.
+ComposeInvestigator is heavily dependent on the version
+of [Compose Compiler](https://developer.android.com/jetpack/androidx/releases/compose-compiler),
+so the version of ComposeInvestigator follows the
+format `[Compose Compiler Version - ComposeInvestigator Version]`.
 
-In other words, you need to adjust the Compose Compiler version and Kotlin version to use ComposeInvestigator.
+In other words, you need to adjust the Compose Compiler version and Kotlin version to use
+ComposeInvestigator.
 
 ---
+
+### Unreleased
+
+- Fixed #123: Tracking `DerivedState` no longer crashes. Instead, due to technical limitations,
+  `DerivedState` is not currently supported for tracking state changes.
 
 ### 1.5.10-0.1.0
 

--- a/gradle.properties
+++ b/gradle.properties
@@ -27,7 +27,7 @@ RELEASE_SIGNING_ENABLED=true
 SONATYPE_AUTOMATIC_RELEASE=true
 
 GROUP=land.sungbin.composeinvestigator
-VERSION_NAME=1.5.10-0.1.0
+VERSION_NAME=1.5.10-0.1.1-SNAPSHOT
 
 POM_DESCRIPTION=Trace the recomposition of a Composable with its cause without any boilerplate code.
 POM_INCEPTION_YEAR=2024

--- a/runtime/src/main/kotlin/land/sungbin/composeinvestigator/runtime/StateObjectTracker.kt
+++ b/runtime/src/main/kotlin/land/sungbin/composeinvestigator/runtime/StateObjectTracker.kt
@@ -7,6 +7,7 @@
 
 package land.sungbin.composeinvestigator.runtime
 
+import android.util.Log
 import androidx.compose.animation.core.Animatable
 import androidx.compose.animation.core.AnimationState
 import androidx.compose.animation.core.InfiniteTransition
@@ -18,7 +19,6 @@ import androidx.compose.runtime.cache
 import androidx.compose.runtime.currentComposer
 import androidx.compose.runtime.snapshots.ObserverHandle
 import androidx.compose.runtime.snapshots.Snapshot
-import androidx.compose.runtime.snapshots.SnapshotMutableState
 import androidx.compose.runtime.snapshots.StateObject
 import land.sungbin.composeinvestigator.runtime.StateObjectTrackManager.stateFieldNameMap
 import land.sungbin.composeinvestigator.runtime.StateObjectTrackManager.stateLocationMap
@@ -118,7 +118,8 @@ internal object StateObjectTrackManager {
  * ```
  *
  * Note that only states where the value of [stateObjectGetter] is not `null`
- * are tracked for value changes.
+ * are tracked for value changes. [androidx.compose.runtime.DerivedState] is
+ * not currently supported.
  *
  * @receiver State to track value changes.
  * @return The same state object.
@@ -151,12 +152,26 @@ public fun <State> State.registerStateObjectTracking(
   val register by lazy {
     object : RememberObserver {
       override fun onRemembered() {
-        val stateObject = stateObjectGetter(state) ?: return
-        trackedStateObjects.getOrPut(composableKeyName, ::mutableSetOf).add(stateObject)
-        stateFieldNameMap.putIfNotPresent(stateObject, stateName)
-        stateValueGetterMap.putIfNotPresent(stateObject, stateValueGetter)
-        stateLocationMap.putIfNotPresent(stateObject, composable)
-        ComposeStateObjectValueGetter.initialize(stateObject)
+        try {
+          val stateObject = stateObjectGetter(state) ?: return
+          trackedStateObjects.getOrPut(composableKeyName, ::mutableSetOf).add(stateObject)
+          stateFieldNameMap.putIfNotPresent(stateObject, stateName)
+          stateValueGetterMap.putIfNotPresent(stateObject, stateValueGetter)
+          stateLocationMap.putIfNotPresent(stateObject, composable)
+          ComposeStateObjectValueGetter.initialize(stateObject)
+        } catch (expectedException: UnsupportedOperationException) {
+          Log.e(
+            ComposeInvestigatorConfig.LOGGER_DEFAULT_TAG,
+            "Failed to execute stateValueGetter from the provided StateObject. Please report this as a project issue.",
+            expectedException,
+          )
+        } catch (unexpectedException: Exception) {
+          Log.e(
+            ComposeInvestigatorConfig.LOGGER_DEFAULT_TAG,
+            "State value tracking registration failed. Please report this as a project issue.",
+            unexpectedException,
+          )
+        }
       }
 
       override fun onForgotten() {
@@ -231,7 +246,7 @@ public object ComposeStateObjectValueGetter : StateValueGetter {
 
   // This may be a pointless defense, but we disable read observers for reliability.
   private fun StateObject.getCurrentValue() = Snapshot.withoutReadObservation {
-    (this as SnapshotMutableState<*>).value
+    (this as? State<*>)?.value ?: throw UnsupportedOperationException("Unsupported StateObject type: ${this::class.java}")
   }
 
   internal fun initialize(key: StateObject) {

--- a/runtime/src/test/kotlin/land/sungbin/composeinvestigator/runtime/StateObjectTrackerTest.kt
+++ b/runtime/src/test/kotlin/land/sungbin/composeinvestigator/runtime/StateObjectTrackerTest.kt
@@ -14,9 +14,11 @@ import androidx.compose.runtime.RecomposeScope
 import androidx.compose.runtime.SideEffect
 import androidx.compose.runtime.currentComposer
 import androidx.compose.runtime.currentRecomposeScope
+import androidx.compose.runtime.derivedStateOf
 import androidx.compose.runtime.getValue
 import androidx.compose.runtime.mutableFloatStateOf
 import androidx.compose.runtime.mutableIntStateOf
+import androidx.compose.runtime.mutableLongStateOf
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.setValue
 import io.kotest.assertions.withClue
@@ -56,7 +58,7 @@ class StateObjectTrackerTest : ShouldSpec() {
       ComposeStateObjectValueGetter.clear()
     }
 
-    should("Receive a callback when the state changes") {
+    should("Receive a callback when the State changes") {
       compositionTest {
         val stringState = mutableStateOf("string")
         val intState = mutableIntStateOf(0)
@@ -115,6 +117,38 @@ class StateObjectTrackerTest : ShouldSpec() {
             STATE(composable = AffectedComposable("floatState"), name = "floatState", previousValue = 1f, newValue = 2f),
           )
         }
+      }
+    }
+    // TODO: Changes in DerivedState are not caught by Snapshot.registerApplyObserver,
+    //  so it doesn't detect when they've changed. To resolve the #123 bug, only
+    //  test that no exceptions are thrown for now.
+    should("No exception is thrown even if DerivedState is present") {
+      compositionTest {
+        val firstNumber = mutableIntStateOf(0)
+        val secondNumber = mutableLongStateOf(1000)
+        val summedNumber = derivedStateOf { firstNumber.intValue + secondNumber.longValue }
+
+        var recomposeScope: RecomposeScope? = null
+        var runCount = 0
+
+        compose {
+          recomposeScope = currentRecomposeScope
+
+          summedNumber.registerStateObjectTracking(
+            composer = currentComposer,
+            composable = AffectedComposable("TEST"),
+            composableKeyName = "KeyName",
+            stateName = "summedNumber",
+          )
+
+          SideEffect {
+            firstNumber.intValue++
+            if (runCount++ > 0) secondNumber.longValue++
+          }
+        }
+
+        recomposeScope!!.invalidate()
+        expectChanges()
       }
     }
     should("Registered callbacks are released when the composable reaches the Forgotten lifecycle") {

